### PR TITLE
Feature: Movie/tv-show details page rating.

### PR DESCRIPTION
**Pull request by Heliken**
localhost_4200_movie_948713

Deployed on [GithubPages](https://heliken.github.io/themoviedb/).
Guest user can now rate movie/tvshow from the detailed page of movie/tv-show, to achieve this functionality:
1. `NotificationService` was added to show response of rating POST method for user;
2. `<detailed-page-hero-rating>` component was updated to show how user rated movie/tv-show + show `<rating-picker>` to submit rating;
3. `<notifications>` component was added to show notifications that are called in `NotificationsService`;
4. `<rating-picker>` component was added to let user select and submit rating for movie/tv-show;
5. `mapGet` pipe was added to pick value from `Map` of stored user ratings;
6. `guestSessionService` was added to let user get guestSessionId and to store it in case there is reason to;
7. `localStorageService` was updated so that expiration timestamp can be optional, as we don't need timestamp for guestSessionid.
8. `ratingService` was added to deal with rating-related data, such as posting guest user rating and requesting guest user submited ratings from api + storing them to have access to values.

### CODE REVIEW

File: src/app/components/notifications/notifications.component.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  constructor(private notificationsService: NotificationsService) {}
+
+  public notifications: CustomNotification[] = [];
+  public notificationsSubscription?: Subscription;
+
+  private notificationDelayTime = 3000;
+
+  ngOnInit(): void {
+    this.notificationsSubscription =
+      this.notificationsService.notifications$.subscribe(notification => {
```
Comment (ItNoN) [inline src/app/components/notifications/notifications.component.ts]: nit: 

```
this.notificationsService.notifications$.pipe(
  tap((notification) => this.notifications.push(notification)),
  delay(this.notificationDelayTime),
  tap(() => this.notifications.shift())
).subscribe();
```

File: src/app/components/notifications/notifications.module.ts [hunk 2]
```diff
@@ -0,0 +1,10 @@
+import { NgModule } from '@angular/core';
+import { CommonModule } from '@angular/common';
+import { NotificationsComponent } from './notifications.component';
+
+@NgModule({
+  declarations: [NotificationsComponent],
+  exports: [NotificationsComponent],
+  imports: [CommonModule],
```
Comment (ItNoN) [inline src/app/components/notifications/notifications.module.ts:8]: nit: remove CommonModule

File: src/app/components/detailed-page/components/detailed-page-hero-rating/detailed-page-hero-rating.component.ts [hunk 3]
```diff
@@ -1,11 +1,23 @@
...
 })
 export class DetailedPageHeroRatingComponent {
+  constructor(private ratingService: RatingService) {}
   @Input() public value = 0;
   @Input() public title = 'Rating: ';
+  @Input() public ratingPostConfig?: RatingPostConfig;
+
+  public showDropdown = false;
+
+  public ratedMaps = this.ratingService.ratedMaps;
```
Comment (ItNoN) [inline src/app/components/detailed-page/components/detailed-page-hero-rating/detailed-page-hero-rating.component.ts]: Why do you need whole ratedMaps? Maybe just ask for certain rating

File: src/app/components/rating-picker/rating-picker.component.ts [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+  public ratingValue = new FormControl(this.initial);
+
+  public submitRating() {
+    if (this.ratingPostConfig && this.ratingValue.value) {
+      const { id, type } = this.ratingPostConfig;
+      const value = this.ratingValue.value;
+
+      this.isSubmitting$.next(true);
+
+      this.guestSessionService
```
Comment (ItNoN) [inline src/app/components/rating-picker/rating-picker.component.ts]: It seems, this component shouldn't know anything about guest session and making request. Maybe we could provide this information from parent components?

File: src/app/services/media-to-grid-media.service.ts [hunk 5]
```diff
@@ -5,7 +5,7 @@ import { Movie } from '../types/movie';
 import { Person } from '../types/person';
 import { TvShow } from '../types/tv-show';
 import { Cast } from '../types/credits';
-import { isCast } from 'app/guards/cast-type-guard';
+import { isCast } from 'src/app/guards/cast-type-guard';
```
Comment (ItNoN) [inline src/app/services/media-to-grid-media.service.ts:8]: It's a bit odd

File: src/app/app.module.ts [hunk 6]
```diff
@@ -9,17 +9,44 @@ import { ApiConfigurationService } from './services/api-configuration.service';
 import { MovieDetailsModule } from './pages/movie-details/movie-details.module';
 import { HeaderModule } from './components/header/header.module';
 import { TvDetailsModule } from './pages/tv-details/tv-details.module';
+import { GuestSessionService } from './services/guest-session.service';
+import { RatingService } from './services/rating.service';
+import { forkJoin, of } from 'rxjs';
+import { NotificationModule } from './components/notifications/notifications.module';
+
+const ratingInitializerFactory =
```
Comment (ItNoN) [inline src/app/app.module.ts]: Maybe you can move this functions to /factory folder somewhere

File: src/app/services/media-to-grid-media.service.ts (hunk 5 above)
Comment (Heliken) [reply src/app/services/media-to-grid-media.service.ts:8]: Don't know how it happend, was shocked aswell.

File: src/app/components/notifications/notifications.module.ts (hunk 2 above)
Comment (Heliken) [reply src/app/components/notifications/notifications.module.ts:8]: Isn't ngFor provided by CommonModule?

Comment (ItNoN) [reply src/app/components/notifications/notifications.module.ts:8]: Sorry, check others module

File: src/app/components/rating-picker/rating-picker.component.ts (hunk 4 above)
Comment (ItNoN) [reply src/app/components/rating-picker/rating-picker.component.ts]: + unsubscribe

Comment (ItNoN) [reply src/app/components/rating-picker/rating-picker.component.ts]: and save session after post rating

Comment (Heliken) [reply src/app/components/rating-picker/rating-picker.component.ts]: Updated: sessionId stored in localStorage only after successful rating POST + unsubscribed in onDestroy hook.

File: src/app/app.module.ts (hunk 6 above)
Comment (Heliken) [reply src/app/app.module.ts]: Thanks,  moved.

File: src/app/components/notifications/notifications.component.ts (hunk 1 above)
Comment (Heliken) [reply src/app/components/notifications/notifications.component.ts]: Updated, thanks.

File: src/app/components/detailed-page/components/detailed-page-hero-rating/detailed-page-hero-rating.component.ts (hunk 3 above)
Comment (Heliken) [reply src/app/components/detailed-page/components/detailed-page-hero-rating/detailed-page-hero-rating.component.ts]: Updated, added getRating method to RatingService.

File: src/app/components/notifications/notifications.module.ts (hunk 2 above)
Comment (Heliken) [reply src/app/components/notifications/notifications.module.ts:8]: Removed CommonModule from modules where it isn't used.

**Review by ItNoN — APPROVED** [review]

**Pull request merged**